util/lru/lru.go

```diff
@@ -91,34 +94,75 @@ func (c *Cache[K, V]) GetOk(key K) (value V, ok bool) {
 // LRU. This should mostly be used for non-intrusive debug inspection
 // of the cache.
 func (c *Cache[K, V]) PeekOk(key K) (value V, ok bool) {
-	if ele, hit := c.m[key]; hit {
-		return ele.Value.(*entry[K, V]).value, true
+	if ent, hit := c.m[key]; hit {
+		return ent.value, true
 	}
 	var zero V
 	return zero, false
 }
 
 // Delete removes the provided key from the cache if it was present.
 func (c *Cache[K, V]) Delete(key K) {
-	if e, ok := c.m[key]; ok {
-		c.deleteElement(e)
+	if ent, ok := c.m[key]; ok {
+		c.deleteElement(ent)
 	}
 }
 
 // DeleteOldest removes the item from the cache that was least recently
 // accessed. It is a no-op if the cache is empty.
 func (c *Cache[K, V]) DeleteOldest() {
-	if c.ll != nil {
-		if e := c.ll.Back(); e != nil {
-			c.deleteElement(e)
-		}
+	if c.l != nil {
+		c.deleteOldest()
 	}
 }
 
-func (c *Cache[K, V]) deleteElement(e *list.Element) {
-	c.ll.Remove(e)
-	delete(c.m, e.Value.(*entry[K, V]).key)
-}
-
 // Len returns the number of items in the cache.
 func (c *Cache[K, V]) Len() int { return len(c.m) }
+
+// newAtFront creates a new LRU entry using key and value, and inserts
+// it at the front of c.l.
+func (c *Cache[K, V]) newAtFront(key K, value V) *entry[K, V] {
+	ret := &entry[K, V]{key: key, value: value}
+	if c.l == nil {
+		ret.prev = ret
+		ret.next = ret
+	} else {
+		ret.next = c.l
+		ret.prev = c.l.prev
+		c.l.prev.next = ret
+		c.l.prev = ret
+	}
+	c.l = ret
+	return ret
+}
+
+// moveToFront moves ent, which must be an element of c.l, to the
+// front of c.l.
+func (c *Cache[K, V]) moveToFront(ent *entry[K, V]) {
+	if c.l == ent {
+		return
+	}
+	ent.prev.next = ent.next
+	ent.next.prev = ent.prev
+	ent.prev = c.l.prev
+	ent.next = c.l
+	c.l.prev.next = ent
+	c.l.prev = ent
+	c.l = ent
+}
+
+// deleteOldest removes the oldest entry in the cache. Panics if there
+// are no entries in the cache.
+func (c *Cache[K, V]) deleteOldest() { c.deleteElement(c.l.prev) }
+
+// deleteElement removes ent, which must be an element of c.l, from
```

Comment:
same: c.l or c.m?

I guess it's saying "already in the ring pointed to by c.l"? Maybe just say "which must already be in the LRU"?

Reply:
Yeah, good call, updated.